src/cactus_client/action/der.py

```diff
@@ -0,0 +1,284 @@
+
+from cactus_test_definitions.csipaus import CSIPAusResource
+
+from cactus_client.action.server import (
+    client_error_request_for_step,
+    resource_to_sep2_xml,
+    submit_and_refetch_resource_for_step,
+)
+from cactus_client.error import CactusClientException
+from cactus_client.model.context import ExecutionContext
+from cactus_client.model.execution import ActionResult, StepExecution
+from envoy_schema.server.schema.sep2.der import (
+    DER,
+    DERCapability,
+    DERType,
+    ActivePower,
+    DERSettings,
+    DERStatus,
+    DERControlType,
+    DOESupportedMode,
+    ConnectStatusTypeValue,
+    OperationalModeStatusTypeValue,
+    OperationalModeStatusType,
+)
+
+from cactus_client.schema.validator import to_hex_binary
+from cactus_client.time import utc_now
+
+logger = logging.getLogger(__name__)
+
+
+def _validate_fields(expected: dict[str, Any], actual: Any) -> None:
+    """Validate that resource fields match expected values. Build the error explanations if not
+
+    Args:
+        expected: Dict of field_name -> expected_value.
+                  Fields with None values are skipped (useful for optional fields).
+        actual: The actual resource object to validate
+    """
+    mismatches = []
+
+    for field_name, expected_value in expected.items():
+        if expected_value is None:
+            continue
+
+        actual_value = getattr(actual, field_name, None)
+
+        if actual_value != expected_value:
+            mismatches.append(f"{field_name} expected {expected_value}, received {actual_value}")
+
+    if mismatches:
+        raise CactusClientException(
+            f"Expected all {actual.__class__.__name__} fields to match. " + "; ".join(mismatches) + "."
+        )
+
+
+async def action_upsert_der_capability(
+    resolved_parameters: dict[str, Any], step: StepExecution, context: ExecutionContext
+) -> ActionResult:
+
+    resource_store = context.discovered_resources(step)
+
+    # Extract and convert parameters
+    type_ = DERType(int(resolved_parameters["type"]))
+    rtgMaxW = ActivePower(value=resolved_parameters["rtgMaxW"], multiplier=0)
+    modesSupported = to_hex_binary(int(resolved_parameters["modesSupported"]), 32)
+    doeModesSupported = to_hex_binary(int(resolved_parameters["doeModesSupported"]), 8)
+
+    # Loop through and upsert the resource for EVERY device
+    stored_der = [sr for sr in resource_store.get(CSIPAusResource.DER)]
+    for der in stored_der:
+        dcap_link = cast(DER, der.resource).DERCapabilityLink
+
+        if dcap_link is None:
+            raise CactusClientException(
+                f"Expected every der to have a DERCapabilityLink, but didnt find one for device {der.resource.href}."
+            )
+
+        # Build the upsert request
+        dcap_request = DERCapability(
+            type_=type_, rtgMaxW=rtgMaxW, modesSupported=modesSupported, doeModesSupported=doeModesSupported
+        )
+        dcap_xml = resource_to_sep2_xml(dcap_request)
```

Comment:
Nitpick - I'd rename dcap_request to dercap_request as dcap is used to describe a DeviceCapability request